assembly/src/parsers/u32_ops.rs

```diff
@@ -882,60 +921,67 @@ fn handle_u32_and_unsafe_check(
+///     bits
+///   - Single argument: any number argument gets pushed to the stack, skips the assert check and
+///     directly performs the operation, dropping high bits
+/// - "Overflowing":
+///   - Zero argument: skips the assert check and directly performs the operation
+///   - Single argument: any number argument gets pushed to the stack, skips the assert check and
+///     directly performs the operation
 ///
 /// According to the spec this is currently U32ADD, U32SUB, U32MUL.
 fn handle_arithmetic_operation(
     span_ops: &mut Vec<Operation>,
     op: &Token,
     arithmetic_op: Operation,
+    op_mode: U32OpMode,
 ) -> Result<(), AssemblyError> {
-    // prepare the stack for the operation and determine if we need to check for overflow
-    let assert_u32_result = match op.num_parts() {
-        0 => return Err(AssemblyError::missing_param(op)),
-        1 => {
-            // for simple arithmetic operation we need to make sure operands are u32 values,
-            // and we need to make sure that the result will be a u32 value as well
-            span_ops.push(Operation::U32assert2);
-            true
-        }
-        2 => match op.parts()[1] {
-            "unsafe" => false,
-            "full" => {
-                // for the full  mode we need to make sure operands are u32 values, but we don't
-                // need to check the result for overflow because we return both high and low bits
-                // of the result
+    let mut drop_high_bits_flag = false;
+    let mut assert_u32_res_flag = false;
+    let num_parts = op.num_parts();
+
+    match op_mode {
+        U32OpMode::Checked => {
+            if num_parts == 1 {
                 span_ops.push(Operation::U32assert2);
-                false
-            }
-            _ => {
-                // for operation.n (where n is the immediate value), we need to push the immediate
-                // value onto the stack, and make sure both operands are u32 values. we also want
-                // to make sure the result is a u32 value.
+            } else {
                 assert_and_push_u32_param(span_ops, op, 0)?;
-                true
             }
-        },
-        _ => return Err(AssemblyError::extra_param(op)),
-    };
+            assert_u32_res_flag = true;
+        }
+        U32OpMode::Wrapping => {
+            if num_parts == 2 {
+                push_param(span_ops, op)?;
+            }
+            drop_high_bits_flag = true;
+        }
+        U32OpMode::Overflowing => {
+            if num_parts == 2 {
+                push_param(span_ops, op)?;
+            }
+        }
+        _ => return Err(AssemblyError::invalid_op(op)),
+    }
 
-    // perform the operation
     span_ops.push(arithmetic_op);
 
-    // make sure the result is a u32 value, and drop the high bits
-    if assert_u32_result {
+    if assert_u32_res_flag {
         span_ops.push(Operation::Eqz);
         span_ops.push(Operation::Assert);
     }
+    if drop_high_bits_flag {
```

Comment:
minor nit: maybe this should be `else if`? I don't think ti should be possible for both of these to be set at the same time.